### PR TITLE
The gleep/update UI rotation speed

### PR DESCRIPTION
**Pull request by Earl-Brown**
right-clicking in the preview area now brings up an options dialog with orbit speed sensitivity and a checkbox for whether or not to display the grid.

Changes made:
* Added UIOptions type
* Added UIOptionsPopup view
* Moved ValueCheckbox and ValueSlider out to Controls folder parallel to App to make them available to UIOptionsPopup
* Renamed "Controls" to "OrbitView" to make it's purpose more clear
* Added sliderColumns and spinnerColumns properties to ValueSlider (it was not using all columns in the Options dialog)
*

### CODE REVIEW

**Comment by Earl-Brown** [comment]
I am *sure* there will be suggestions for formatting and code style - I welcome them, and will keep an eye out to make the changes as soon as possible!

File: src/views/UIOptionsPopup/UIOptionsPopup.tsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+const UIOptionsPopup: React.FC<Props> = ({ children }: Props) => {
+  const [isVisible, setVisibility] = useGlobalState('isUIOptionsPopupVisible')
+
+  const showPopup = () => setVisibility(true)
+  const hidePopup = () => setVisibility(false)
+
+  const optionsDialog = isVisible ? (
+    <UIOptionsDialog>
```
Comment (accassid) [inline src/views/UIOptionsPopup/UIOptionsPopup.tsx]: I like the tenacity to create a custom dialog popup. However for a project like this where we already have UI library installed, it's generally better to just use a component already available to you rather than making one from scratch. It both reduces code overhead and adds to UI uniformity on the app. I'd recommend using the antd modal for this one. Check out the docs here [https://ant.design/components/modal/](https://ant.design/components/modal/)

File: src/views/UIOptionsPopup/UIOptionsPopup.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      <div style={{ margin: '2px 8px', height: '0.9em', top: '-0.6em', position: 'relative' }}>
+        <ValueCheckbox stateKey={'showGrid'} label="Show Grid" />
+      </div>
+    </UIOptionsDialog>
+  ) : (
+    <></>
+  )
+
+  return (
+    <div onContextMenu={showPopup} style={{ height: 'inherit', width: 'inherit}' }}>
```
Comment (accassid) [inline src/views/UIOptionsPopup/UIOptionsPopup.tsx]: Interesting thought putting it on a context menu trigger, however I don't think that's necessary for this. First off, people wouldn't intuitively know to do that (I had to look at the code to figure it out) and second people often get cranky when they have their regular context menus overridden in the browser. To that end, I would just use a settings icon from [antd](https://ant.design/components/icon/) and make a button out of it that triggers the menu. I'd put that button either in the toolbar by the menu button, in the upper right of the settings panel, or maybe just make it another button with the social media. Play around and see what looks best.

File: src/models/UIOptions.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+export type UIOptions = {
+  orbitSpeed: number
+  showGrid: number
+  isUIOptionsPopupVisible: true | false
```
Comment (accassid) [inline src/models/UIOptions.ts]: I'd probably just use the `boolean` type here.

File: src/views/PageWrapper/PageWrapper.tsx [hunk 4]
```diff
@@ -26,7 +28,7 @@ const PageWrapper: React.FC<Props> = ({ children }: Props) => {
       <TopToolbar />
       <PageBody>
         <RightPanel />
-        {children}
+        <UIOptionsPopup>{children}</UIOptionsPopup>
```
Comment (accassid) [inline src/views/PageWrapper/PageWrapper.tsx]: If you use a button triggered modal, as suggested later, no need to pass the page body children into this component for the context menu trigger.

**Review by accassid — CHANGES_REQUESTED** [review]
Overall, good first PR. Your types, models, and file placements all seem reasonable to me. The big thing I'd like to see changed is the styling of the menu itself and its trigger as I commented. Nice work.

**Comment by Earl-Brown** [comment]
Thanks for the feedback - I will work to make these changes.

File: src/views/UIOptionsPopup/UIOptionsPopup.tsx (hunk 1 above)
Comment (Earl-Brown) [reply src/views/UIOptionsPopup/UIOptionsPopup.tsx]: TBH - I was having trouble getting the modal to work ... but I've learned a bit, so it'll probably go easier when I adapt to all of your suggestions :)

**Comment by Earl-Brown** [comment]
I have revised the interface to honor suggestions:
* changed menu to icon so it indicates what will happen when it is clicked (collapsed or expanded)
* Added a "config" icon
* bound the configuration dialog to the config button and changed where it appears

File: src/views/PageWrapper/PageWrapper.tsx (hunk 4 above)
Comment (Earl-Brown) [reply src/views/PageWrapper/PageWrapper.tsx]: I have submitted changes to honor your suggestions!

File: src/models/UIOptions.ts (hunk 3 above)
Comment (Earl-Brown) [reply src/models/UIOptions.ts]: I have submitted changes to honor your suggestions!

File: src/views/UIOptionsPopup/UIOptionsPopup.tsx (hunk 2 above)
Comment (Earl-Brown) [reply src/views/UIOptionsPopup/UIOptionsPopup.tsx]: I added a configuration button, and removed the context menu trigger.

File: src/views/UIOptionsPopup/UIOptionsPopup.tsx (hunk 1 above)
Comment (Earl-Brown) [reply src/views/UIOptionsPopup/UIOptionsPopup.tsx]: I have submitted changes to honor your suggestions!

**Comment by Earl-Brown** [comment]
In my earlier comments, I forgot to mention that I'm leaving my original popup instead of using a modal, because I couldn't see a way to have a modal allow the user to interact with the preview; I like it better to have the users able to test the speed they've set before they close the window.

File: src/views/UIOptionsPopup/ConfigurationDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  const togglePopup = () => {
+    setVisibility(!isVisible)
+  }
+
+  return (
+    <>
+      <SettingOutlined onClick={togglePopup} style={style}></SettingOutlined>
+      {isVisible ? (
+        <UIOptionsDialog>
+          <h3 style={{ borderBottom: '1px solid white' }}>Settings</h3>
```
Comment (accassid) [inline src/views/UIOptionsPopup/ConfigurationDialog.tsx]: I would change this line color to match the border color of your UIOptionsDialog.

File: src/views/UIOptionsPopup/style.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import styled from 'styled-components'
+
+export const UIOptionsDialog = styled.div`
+  top: 40px;
+  right: 16px;
+  width: 38em;
+  position: absolute;
+  background-color: rgba(0, 0, 0, 0.9);
```
Comment (accassid) [inline src/views/UIOptionsPopup/style.ts]: To contribute to the app's uniformity, I'd change this background and opacity to match that of the RightPanel's. I believe that should be `rgba(28, 28, 28, 0.9)`.

File: src/views/UIOptionsPopup/style.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+  background-opacity: 0.5;
```
Comment (accassid) [inline src/views/UIOptionsPopup/style.ts]: Shouldn't need this property if you're using rgba.

File: src/views/UIOptionsPopup/style.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  border: 2px solid silver;
```
Comment (accassid) [inline src/views/UIOptionsPopup/style.ts]: More uniformity points, I'd use the same grey that the UI library is using, so change `silver` to `#434343`

File: src/views/UIOptionsPopup/ConfigurationDialog.tsx (hunk 5 above)
Comment (accassid) [reply src/views/UIOptionsPopup/ConfigurationDialog.tsx]: See my color suggestion for the border in the style file.

File: src/views/TopToolbar/TopToolbar.tsx [hunk 9]
```diff
@@ -36,7 +39,12 @@ const TopToolbar: React.FC<Props> = () => {
         </a>
       </BarSegment>
       <BarSegment>
-        <MenuOutlined onClick={togglePanel} style={{ fontSize: 25, cursor: 'pointer' }} />
+        <ConfigurationDialog style={{ ...iconStyle, margin: '0 0.5em' }} />
+        {rightPanelActive ? (
```
Comment (accassid) [inline src/views/TopToolbar/TopToolbar.tsx]: I like the concept of what you're doing here, however I'd remove it from this PR for two reasons:
1. It's out of the scope of this PR, so if you think it should be added open a different PR. PRs can be small!
2. This doesn't work with the rightPanelActiveMobile property which is the inverse of the rightPanelActive property, meaning the way you have this coded the icon will be backwards on mobile.

File: src/views/UIOptionsPopup/style.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
+import styled from 'styled-components'
+
+export const UIOptionsDialog = styled.div`
```
Comment (accassid) [inline src/views/UIOptionsPopup/style.ts:3]: I'd add a box shadow to contribute to its floating sense:
`box-shadow: 5px 5px 8px 0px rgba(0,0,0,0.68);`

File: src/views/UIOptionsPopup/style.ts [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+  top: 40px;
```
Comment (accassid) [inline src/views/UIOptionsPopup/style.ts]: I'd make this 43px to give it more of a floating sense.

File: src/views/UIOptionsPopup/style.ts (hunk 10 above)
Comment (accassid) [reply src/views/UIOptionsPopup/style.ts:3]: Also, make sure to always check UI elements in responsive mode to make sure they work on mobile. This element looks like this right now:
Screenshot_2021-02-24_12-12-40
You should be able to solve this by adding a `max-width: 95%` property. However, you'll have to correct some issues with the ValueSlider overflowing then.

**Review by accassid — CHANGES_REQUESTED** [review]
I can see what you mean that you want to be able to test the responsiveness while the popout is open, so I'll allow it. That being said, I did request some styling changes. Also you have some warnings to resolve. Make sure before a PR that in the terminal where you ran `yarn start` there are no active warnings.
Screenshot from 2021-02-23 14-48-39

File: src/views/UIOptionsPopup/ConfigurationDialog.tsx (hunk 5 above)
Comment (Earl-Brown) [reply src/views/UIOptionsPopup/ConfigurationDialog.tsx]: Will do

File: src/views/UIOptionsPopup/style.ts (hunk 6 above)
Comment (Earl-Brown) [reply src/views/UIOptionsPopup/style.ts]: Thanks for the suggestion!  I wasn't sure what to use.

File: src/views/UIOptionsPopup/style.ts (hunk 7 above)
Comment (Earl-Brown) [reply src/views/UIOptionsPopup/style.ts]: Again, thanks!

File: src/views/UIOptionsPopup/style.ts (hunk 8 above)
Comment (Earl-Brown) [reply src/views/UIOptionsPopup/style.ts]: >thumbup<

File: src/views/TopToolbar/TopToolbar.tsx (hunk 9 above)
Comment (Earl-Brown) [reply src/views/TopToolbar/TopToolbar.tsx]: OK - I'll pull it for now.

**Comment by Earl-Brown** [comment]
Thanks for all of the comments and suggestions!

I'll work to fix them.  Be prepared to find more issues :)

And I'll fix them as soon as I can!

File: src/views/UIOptionsPopup/style.ts (hunk 10 above)
Comment (Earl-Brown) [reply src/views/UIOptionsPopup/style.ts:3]: Done.  No overflow problem at all!

File: src/views/UIOptionsPopup/ConfigurationDialog.tsx (hunk 5 above)
Comment (Earl-Brown) [reply src/views/UIOptionsPopup/ConfigurationDialog.tsx]: Made the color changes and fixed the compile warnings.

File: src/views/UIOptionsPopup/style.ts (hunk 6 above)
Comment (Earl-Brown) [reply src/views/UIOptionsPopup/style.ts]: Done

File: src/views/UIOptionsPopup/style.ts (hunk 7 above)
Comment (Earl-Brown) [reply src/views/UIOptionsPopup/style.ts]: Done

File: src/views/UIOptionsPopup/style.ts (hunk 8 above)
Comment (Earl-Brown) [reply src/views/UIOptionsPopup/style.ts]: Done

File: src/views/TopToolbar/TopToolbar.tsx (hunk 9 above)
Comment (Earl-Brown) [reply src/views/TopToolbar/TopToolbar.tsx]: Done

File: src/views/UIOptionsPopup/style.ts (hunk 11 above)
Comment (Earl-Brown) [reply src/views/UIOptionsPopup/style.ts]: Done

**Comment by accassid** [comment]
Looks good! Seems to be just one compilation warning. Fix that up and it should be good to go. 
Screenshot from 2021-03-08 18-08-57

**Comment by accassid** [comment]
> Looks good! Seems to be just one compilation warning. Fix that up and it should be good to go.
> Screenshot from 2021-03-08 18-08-57

Actually, nevermind, that must have been on dev when you forked it and it's fixed now. Never bad to rebase before a PR though.

**Comment by Earl-Brown** [comment]
Sweet,  Thanks! 

On March 8, 2021 6:14:56 PM MST, Andrew Cassidy <notifications@github.com> wrote:
>Merged #88 into develop.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/accassid/dice-gen/pull/88#event-4424507764

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.
